### PR TITLE
SurfaceView Maintenance

### PR DESCRIPTION
**Pull request by Scrappers-glitch**
A maintenance PR for the `JmeSurfaceView`.

## New features introduced : 

1) Optionally  binding the jmeSurfaceView to the android activity life cycle, it helps in `pausing`, `resuming` game and openGL context, and also helps in releasing memory when the game is destroyed.

2) Added new Listeners events : `OnLayoutDrawn()`, `OnRendererStarted()`.

4) Added a `DestructionPolicy` flag, which uses a static game state `GameState` to determine whether to destroy the game with the activity on exiting the activity or to keep it for a second use.
https://youtu.be/NBWyBzr_YXU

5) Optimizations to the `ErrorDialog` includes : showing a full error stack and `Copy crash log` button.

6) Added other flags such as : `showErrorDialog`, `exitOnEscPressed`, `showEscExitPrompt`.

7) Added a getter field for gl_es version.

8) Documentations and code optimizations following best practices.

`NB : changes here are tested more than once on pre-existing jme3 android apps since jme3.4.0, and they won't likely affect user code.`

### Screenshots : 

Copy crash log feature

I will try to make a video to preview these changes on a live android test.

### CODE REVIEW

**Comment by stephengold** [comment]
Thanks for your contribution to the Engine!

File: jme3-android/src/main/java/com/jme3/app/jmeSurfaceView/JmeSurfaceView.java [hunk 1]
```diff
@@ -80,209 +91,411 @@
...
+     * Adds the glSurfaceView to the screen immediately, saving the current app instance.
+     */
+    protected void addGlSurfaceView(){
+        /*jme Renderer joins the UIThread at that point*/
+        JmeSurfaceView.this.addView(glSurfaceView);
+        /*dispatch the layout drawn event*/
+        if(onLayoutDrawn != null){
+            onLayoutDrawn.onLayoutDrawn(legacyApplication, this);
+        }
+        /*set the static memory to hold the game state, only if the game life cycle is not bounded to the activity life cycle*/
```
Comment (Scrappers-glitch) [inline jme3-android/src/main/java/com/jme3/app/jmeSurfaceView/JmeSurfaceView.java:314]: Comment should be changed to `Only if the destruction policy uses KEEP_WHEN_FINISHED policy`
In either polices, the user can choose if he/she wants to keep the surface view updated with a activity life cycle or not !

I will correct this.

Comment (Scrappers-glitch) [reply jme3-android/src/main/java/com/jme3/app/jmeSurfaceView/JmeSurfaceView.java:314]: This comment is wrongly added before the think of adding a destruction policy to control this.

**Comment by Scrappers-glitch** [comment]
I will review this PR again today to finally assure the changes.

**Comment by stephengold** [comment]
I'd feel more comfortable with this if it were reviewed by someone other than the author. I can't give good feedback because I'm not familiar with Android.

Unless there's substantial review, I intend to integrate this in 24 hours.

**Comment by Scrappers-glitch** [comment]
> I'd feel more comfortable with this if it were reviewed by someone other than the author. I can't give good feedback because I'm not familiar with Android.
> 
> Unless there's substantial review, I intend to integrate this in 24 hours.

You can give it a quick look, if you find something that is odd, point it out and i will try to modify it to be as friendly and good as i can.

**Comment by stephengold** [comment]
Like I said, I can't give this a proper review.

One thing I'm curious about is the addition of `enum GameState` with no values. It looks to me a utility class, but I've never seen a utility class defined as an `enum`. Why did you do it that way?

**Comment by stephengold** [comment]
Also, if this isn't ready to be integrated, please mark it as a "draft" so it won't get integrated prematurely.

**Comment by Scrappers-glitch** [comment]
> Like I said, I can't give this a proper review.
> 
> One thing I'm curious about is the addition of `enum GameState` with no values. It looks to me a utility class, but I've never seen a utility class defined as an `enum`. Why did you do it that way?

A shortcut to the utility class instead of creating a class and an object, if you don't like it or feeling its off, I would convert it to a static subclass.

**Comment by stephengold** [comment]
> A shortcut to the utility class

I've no idea whether this is an acceptable idiom or not.

**Comment by Scrappers-glitch** [comment]
> > A shortcut to the utility class
> 
> I've no idea whether this is an acceptable idiom or not.

I see, but the enum here acts as just a scope, so instead of saying `setLegacyApplication()` directly, we say `GameState.setLegacyApplication()` which keeps our game instance as a static state.

**Comment by pspeed42** [comment]
I've never seen anyone use an enum to hack the singleton pattern.  It's weird and I don't recommend it.

What advantage does this have over a static method on a regular class?  Other than being confusing?

**Comment by Scrappers-glitch** [comment]
> I've never seen anyone use an enum to hack the singleton pattern. It's weird and I don't recommend it.
> 
> What advantage does this have over a static method on a regular class? Other than being confusing?

Alright, I will use a static subclass instead.

**Comment by Scrappers-glitch** [comment]
It's now ready to be merged, i have converted the `enum` to `static subclass` and add the missing constructor and some more clarifications to the constructors, those constructors are default styling constructors.

**Comment by stephengold** [comment]
You said this PR was done, and I integrated it. Why are you still pushing commits to it?

**Comment by Scrappers-glitch** [comment]
> You said this PR was done, and I integrated it. Why are you still pushing commits to it?

I have not pushed anything else ?

**Comment by stephengold** [comment]
You're right about the content of the merged commit. Thanks for pointing that out.

Sorry for the confusion.

**Pull request merged**